### PR TITLE
Do not postprocess internal queries in Migrator

### LINKED ISSUE #2644: Undefined max_batch when using knexSnakeCaseWrappers (Objection)
# Bug
When running `knex` migrations configured with Objection's `knexSnakeCaseWrappers`, the promise handler for the `_latestBatchNumber` query **always** returns a default value of `0`:
```node
// knex/src/migrate/index.js, lines 329-332

_latestBatchNumber(trx = this.knex) {
    return trx.from(getTableName(this.config.tableName, this.config.schemaName))
      .max('batch as max_batch').then(obj => obj[0].max_batch || 0);
  }

```

This is because `max_batch` does not exist in `obj`. The actual structure of `obj`, due to `knexSnakeCaseWrappers` is:
```node
obj = {
  maxBatch: 2
}
```

Defaulting to `0` results in a batch number of `1` always being inserted into the `knex_migrations` table when using `knexSnakeCaseWrappers`.

In effect, this makes all of your migrations only use a single batch, which is potentially **_catastrophic_** when performing a rollback.

# Environment

Knex version: 0.14.6
Database + version: Postgres 10.1
OS: macOS/Linux

Select applicable template from below.
If issue is about oracledb support tag @ atiertant. For MSSql tag @ smorey2 .
Rest of dialects doesn't need tags.

### PR DESCRIPTION
**Pull request by kibertoad**
fixes #2644
Also fixes .withUserParams() method to work correctly with .queryBuilder()

### CODE REVIEW

File: src/migrate/table-creator.js [hunk 1]
```diff
@@ -10,21 +10,26 @@ export function ensureTable(tableName, schemaName, trxOrKnex) {
...
     .hasTable(tableName)
-    .then(
-      (exists) =>
-        !exists && _createMigrationTable(tableName, schemaName, trxOrKnex)
-    )
-    .then(() => getSchemaBuilder(trxOrKnex, schemaName).hasTable(lockTable))
-    .then(
-      (exists) =>
+    .then((exists) => {
+      return !exists && _createMigrationTable(tableName, schemaName, trxOrKnex);
```
Comment (kibertoad) [inline src/migrate/table-creator.js]: it's a huge pain to debug arrow functions without {} and prettier really butchers formatting for them, hence the change

File: src/util/make-knex.js [hunk 2]
```diff
@@ -10,24 +10,23 @@ import batchInsert from './batchInsert';
 export default function makeKnex(client) {
   // The object we're potentially using to kick off an initial chain.
   function knex(tableName, options) {
```
Comment (kibertoad) [inline src/util/make-knex.js]: There are more changes here than I am comfortable with, TBH. Main reason - there was actually a creeping bug after withUserParams change - invoking knex('tableName') would silently create a builder from wrong (original) knex instance. While figuring out a solution I've implemented a stricter separation between knex() function (which should be as lightweight as possible) and the actual knex execution context (that knex function proxies to). I'm still not happy about our magical knex entity which is function, object and class instance at the same time, but at least there is clearer separation of what is responsible for what now)

File: test/unit/migrate/migrations/blank_migration.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+exports.up = function(knex) {};
```
Comment (kibertoad) [inline test/unit/migrate/migrations/blank_migration.js]: having this kinda violates the rule of unit tests having no dependencies, but mocking out require() call for a migration or restructuring migrator to be able to mock out entire sequences is a huge pain, and we don't really need an integration test that would be going over all dialects here at all as we are not testing any db functionality here.

File: src/migrate/Migrator.js [hunk 4]
```diff
@@ -46,10 +47,15 @@ const CONFIG_DEFAULT = Object.freeze({
...
 export default class Migrator {
   constructor(knex) {
-    this.knex = knex;
-    this.config = getMergedConfig(knex.client.config.migrations);
-    this.generator = new MigrationGenerator(knex.client.config.migrations);
-
+    // Clone knex instance and remove post-processing that is unnecessary for internal queries from a cloned config
+    this.knex = isFunction(knex)
+      ? knex.withUserParams(knex.userParams)
+      : Object.assign({}, knex);
```
Comment (heisian) [inline src/migrate/Migrator.js]: This is great! I attempted to clone knex but unsuccessfully in some prototyping code - I would receive an error (forgot what it was) attempting to make queries with my cloned version. Good to see how it is done here.

Comment (kibertoad) [reply src/migrate/Migrator.js]: Admittedly I'm abusing .withUserParams function, but given that it includes all the necessary black magic for cloning to actually produce a fully usable knex instance, that's the best option we have now. Maybe we should eventually replace .clone function with what .withUserParams currently does and call it under-the-hood.

**Comment by heisian** [comment]
@elhigu When do you think the next tagged release will be?

**Comment by kibertoad** [comment]
@heisian Soon. I'm updating changelog and package.json today, but I also hope to squeeze https://github.com/tgriesser/knex/pull/2861 in - which depends on external project merging and releasing; if it doesn't happen tomorrow, probably we are going to release without it.

**Comment by heisian** [comment]
@kibertoad dope, thank you

**Comment by dlangerenken** [comment]
just to be sure: Is this also including the wrapIdentifier function for migration that also converts knex_migrations_lock and other internal tables? This does not seem to be fixed in 0.16.3

**Comment by elhigu** [comment]
@dlangerenken Sounds like you might be right. Looks like implementation tries to reset wrapIdentifier to default too, but it does not seem to be tested by https://github.com/tgriesser/knex/pull/2914/files#diff-09c603a78f9c37fe1a94f869d4a83405R16

That test case needs to be extended to cover it too.

**Pull request merged**